bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -65,6 +68,20 @@ object BagsState {
   type StateTransition = (BagsState, IO[Unit]) // Next BagsState, side-effects
   val ErrorTransition  = (ErrorState, ioUnit)  // Default transition for unexpected events
 
+  def hashObs(ctx: ObsContext): AgsHashVal = {
+    val when = ctx.getSchedulingBlockStart.asScalaOpt | Instant.now.toEpochMilli
+    AgsHash.hash(ctx, when)
+  }
+
+  def hashObs(o: ISPObservation): Option[AgsHashVal] =
+    ObsContext.create(o).asScalaOpt.map(hashObs)
+
+  def fetchObs(k: ObsKey): Option[ISPObservation] =
+    Option(SPDB.get().lookupProgram(k.prog.key)).flatMap { p =>
```

Comment:
It's kind of a hack, admittedly.  `SPDB` is assigned with the local ODB on startup.  This is used throughout the `OT` so I don't feel *too* guilty about it.  But I do feel a bit guilty about it.